### PR TITLE
model: don't use --app for relation-get/-set before 2.7.0

### PR DESCRIPTION
**Pull request by chipaca**
Juju before 2.7.0 didn't support the `--app` parameter for
`relation-get` and `relation-set`, meaning that the current code
breaks in 2.6.9.

This fixes that by avoiding `--app` entirely when not needed,
raising an explicit `RuntimeError` when an attempt is made to get
or set app data in an old Juju (where we know adding `--app` will
fail).

fixes: #223

### CODE REVIEW

File: ops/model.py [hunk 1]
```diff
@@ -996,10 +997,15 @@ def relation_get(self, relation_id, member_name, is_app):
         if not isinstance(is_app, bool):
             raise TypeError('is_app parameter to relation_get must be a boolean')
 
+        if is_app and not JujuVersion.from_environ().has_app_data():
+            return {}
+
+        args = ['relation-get', '-r', str(relation_id), '-', member_name]
+        if is_app:
+            args.append('--app')
```
Comment (jameinel) [inline ops/model.py:1005]: I'm happy to use this logic always rather than the '--app=False' logic. It makes it easier to support 2.6.* if-and-when for anything that doesn't need app data.

File: ops/model.py [hunk 2]
```diff
@@ -996,10 +997,15 @@ def relation_get(self, relation_id, member_name, is_app):
         if not isinstance(is_app, bool):
             raise TypeError('is_app parameter to relation_get must be a boolean')
 
+        if is_app and not JujuVersion.from_environ().has_app_data():
+            return {}
```
Comment (jameinel) [inline ops/model.py:1001]: I'm hesitant on this one. Depending on how they're using application data, they may *need* it to be there. So I think I'd rather have a different error that is clearer about 'this isn't supported by Juju' and charms can catch that exception and deal with it.
Silently returning {} is dangerous, IMO.

Also, the bugs in 2.7.* were that you couldn't read your *own* application data. Reading the application data for a related application works fine. (app A related to app B, A/0 can read B's application data, and B/0 can read A's application data, the bug was that the leader A/0 couldn't read A's application data.)

Which means that this would preclude charms from using relation application data even where Juju *did* support it.

File: ops/jujuversion.py [hunk 3]
```diff
@@ -83,3 +84,15 @@ def __lt__(self, other):
...
+    def from_environ(cls) -> 'JujuVersion':
+        """Build a JujuVersion from JUJU_VERSION."""
+        v = os.environ.get('JUJU_VERSION')
+        if not v:
+            raise RuntimeError('environ has no JUJU_VERSION')
+        return cls(v)
+
+    def has_app_data(self) -> bool:
+        """Determine whether this juju version knows about app data."""
+        return (self.major, self.minor, self.patch) >= (2, 7, 1)
```
Comment (jameinel) [inline ops/jujuversion.py:98]: I touched on this briefly but I want to be a bit clearer.
* 2.6.* doesn't have any concept of application data
* 2.7.0 has application data, and 2 different applications can use it to share information between them
* 2.7.0 has a bug when a charm wants to read the application data it is sending to the other side. (you can write it fine, but you can't write it and then read it back.)
* I think there is also a bug about application data in a peer relation, and a bug about application data on Kubernetes, where the unit agents run under a different context (there is an Application Agent that runs each unit logic, vs a separate Unit agent that runs on a different machine.)

**Comment by chipaca** [comment]
I've changed it as discussed, such that it only adds `--app` if needed and supported, and raises an exception when the feature is requested but the Juju does not support it.

IMHO we can look into the more nuanced 2.7.x bugs separately, if needed by our users; I think a bit of hand holding to catch and explain what's going on might be in reasonable there. Alternatively maybe just logging a warning (e.g. directly in `has_app_data`) immediately before attempting something that might fail because of those known bugs might be a good middle ground.

**Comment by chipaca** [comment]
e.g. `app data has known bugs before 2.7.4 (see lp:1854348 and lp:1865229), the following call might fail`. Or something. Idea being it should be little work on our side but still not waste the users' time trying to figure *why* it broke. We can't stop it from breaking :)

File: ops/model.py [hunk 4]
```diff
@@ -996,10 +997,15 @@ def relation_get(self, relation_id, member_name, is_app):
         if not isinstance(is_app, bool):
             raise TypeError('is_app parameter to relation_get must be a boolean')
 
+        if is_app and not JujuVersion.from_environ().has_app_data():
+            raise RuntimeError('trying to relation-get app data on a Juju too old for it')
```
Comment (jameinel) [inline ops/model.py:1001]: what about:
'relation-get of application data is not supported by this version of Juju {}'.format(version)

File: ops/model.py [hunk 5]
```diff
@@ -1009,9 +1015,15 @@ def relation_set(self, relation_id, key, value, is_app):
         if not isinstance(is_app, bool):
             raise TypeError('is_app parameter to relation_set must be a boolean')
 
+        if is_app and not JujuVersion.from_environ().has_app_data():
+            raise RuntimeError('trying to relation-set app data on a Juju too old for it')
```
Comment (jameinel) [inline ops/model.py:1019]: we'd want similar wording here.
I we are exposing this as 'relation-set' then I would probably say 'relation-set --app'.
If we are trying to talk it terms of Python objects, then I would say something more like:
'setting application relation data is not supported on older Juju versions'
or 'on Juju {version}'
or 'on Juju versions older than 2.7.0'

File: test/test_model.py [hunk 6]
```diff
@@ -931,29 +933,79 @@ def test_relation_tool_errors(self):
...
+
+        fake_script(self, 'relation-set', 'exit 0')
+
+        # on 2.7.0+, things proceed as expected
+        for v in ['2.8.0', '2.7.0']:
+            with self.subTest(v):
+                os.environ['JUJU_VERSION'] = v
+                self.backend.relation_set(1, 'foo', 'bar', is_app=True)
+                calls = [' '.join(i) for i in fake_script_calls(self, clear=True)]
+                self.assertEqual(calls, ['relation-set -r 1 foo=bar --app'])
```
Comment (jameinel) [inline test/test_model.py:1002]: If we feel like this is cleaner, can we just have that be the formatted version of fake_script_calls ?
Or is that what you wanted to do in a separate patch? (I think I like it more, as it is easier to write and read the tests and we don't care that sometimes the string may not be unambiguously parsable.)

Comment (chipaca) [reply test/test_model.py:1002]: it wasn't part of what I was looking to refactor, but it does make things more readable for me.
I'll keep it on the backburner :-)

File: ops/model.py (hunk 5 above)
Comment (chipaca) [reply ops/model.py:1019]: I like the second one :-)
maybe `[SG]etting application data is not supported on Juju versions older than 2.7.0 (you have {})`

Comment (chipaca) [reply ops/model.py:1019]: hm, actually mentioning 2.7.0 in the error means we track it in two different places
